File: src/org/elixir_lang/reference/resolver/CallDefinitionClause.java
package org.elixir_lang.reference.resolver;

import com.intellij.openapi.util.Pair;
import com.intellij.psi.PsiElementResolveResult;
import com.intellij.psi.ResolveResult;
import com.intellij.psi.impl.source.resolve.ResolveCache;
import org.apache.commons.lang.math.IntRange;
import org.elixir_lang.psi.call.Call;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.macroChildCalls;
import static org.elixir_lang.structure_view.element.CallDefinitionClause.enclosingModularMacroCall;
import static org.elixir_lang.structure_view.element.CallDefinitionSpecification.typeNameArity;

public class CallDefinitionClause implements ResolveCache.PolyVariantResolver<org.elixir_lang.reference.CallDefinitionClause> {
    public static final CallDefinitionClause INSTANCE = new CallDefinitionClause();

    @NotNull
    private static List<ResolveResult> add(@Nullable List<ResolveResult> resolveResultList,
                                           @NotNull Call call,
                                           boolean validResult) {
        @NotNull List<ResolveResult> nonNullResolveResultList;

        if (resolveResultList != null) {
            nonNullResolveResultList = resolveResultList;
        } else {
            nonNullResolveResultList = new ArrayList<>();
        }

        nonNullResolveResultList.add(new PsiElementResolveResult(call, validResult));

        return nonNullResolveResultList;
    }

    @NotNull
    @Override
    public ResolveResult[] resolve(@NotNull org.elixir_lang.reference.CallDefinitionClause callDefinitionClause,
                                   boolean incompleteCode) {
        Call enclosingModularMacroCall = enclosingModularMacroCall(callDefinitionClause.moduleAttribute);
        List<ResolveResult> resolveResultList = null;

        if (enclosingModularMacroCall != null) {
            Call[] siblings = macroChildCalls(enclosingModularMacroCall);

            if (siblings != null && siblings.length > 0) {
                Pair<String, Integer> nameArity = typeNameArity(callDefinitionClause.getElement());
                String name = nameArity.first;
                int arity = nameArity.second;

                for (Call call : siblings) {
                    if (org.elixir_lang.structure_view.element.CallDefinitionClause.is(call)) {
                        Pair<String, IntRange> callNameArityRange =
                                org.elixir_lang.structure_view.element.CallDefinitionClause.nameArityRange(call);

                        if (callNameArityRange != null) {
                            String callName = callNameArityRange.first;

                            if (callName.equals(name)) {
                                IntRange callArityRange = callNameArityRange.second;

                                if (callArityRange.containsInteger(arity)) {
                                    resolveResultList = add(resolveResultList, call, true);
                                } else if (arity < callArityRange.getMaximumInteger()) {
                                    resolveResultList = add(resolveResultList, call, false);
                                }
                            } else if (incompleteCode && callName.startsWith(name)) {
                                IntRange callArityRange = callNameArityRange.second;

                                if (callArityRange.containsInteger(arity)) {
                                    resolveResultList = add(resolveResultList, call, false);
                                } else if (arity < callArityRange.getMaximumInteger()) {
                                    resolveResultList = add(resolveResultList, call, false);
                                }
                            }
                        }
                    }
                }
            }
        }

        ResolveResult[] resolveResults;

        if (resolveResultList != null) {
            resolveResults = resolveResultList.toArray(new ResolveResult[resolveResultList.size()]);
        } else {
            resolveResults = ResolveResult.EMPTY_ARRAY;
        }

        return resolveResults;
    }
}

File: src/org/elixir_lang/reference/resolver/CallDefinitionClause.kt
package org.elixir_lang.reference.resolver

import com.intellij.psi.PsiElementResolveResult
import com.intellij.psi.ResolveResult
import com.intellij.psi.ResolveState
import com.intellij.psi.impl.source.resolve.ResolveCache
import org.elixir_lang.Arity
import org.elixir_lang.Name
import org.elixir_lang.NameArityInterval
import org.elixir_lang.psi.For
import org.elixir_lang.psi.call.Call
import org.elixir_lang.psi.impl.call.macroChildCalls
import org.elixir_lang.semantic.call.definition.Clause
import org.elixir_lang.semantic.call.definition.Delegation
import org.elixir_lang.semantic.semantic

object CallDefinitionClause : ResolveCache.PolyVariantResolver<org.elixir_lang.reference.CallDefinitionClause> {
    override fun resolve(
        callDefinitionClause: org.elixir_lang.reference.CallDefinitionClause,
        incompleteCode: Boolean
    ): Array<ResolveResult> =
        callDefinitionClause.specification.let { specification ->
            specification
                .enclosingModular
                .psiElement.let { it as? Call }
                ?.macroChildCalls()?.let { siblings ->
                    if (siblings.isNotEmpty()) {
                        val nameArity = specification.nameArity
                        val name = nameArity.name
                        val arity = nameArity.arity

                        siblings
                            .flatMap { call -> callToResolveResults(call, name, arity) }
                            .toTypedArray()
                    } else {
                        null
                    }
                }
                ?: emptyArray()
        }

    private fun callToResolveResults(call: Call, name: Name, arity: Arity): List<ResolveResult> =
        when (val semantic = call.semantic) {
            is Clause -> {
                semantic
                    .nameArityInterval
                    ?.let { nameArityInterval ->
                        nameArityIntervalToResolveResult(
                            call,
                            name,
                            arity,
                            nameArityInterval
                        )
                    }
                    ?.let { listOf(it) }
                    .orEmpty()
            }
            is Delegation -> {
                semantic
                    .nameArityInterval
                    ?.let { nameArityInterval ->
                        nameArityIntervalToResolveResult(
                            call,
                            name,
                            arity,
                            nameArityInterval
                        )
                    }
                    ?.let(::listOf)
                    .orEmpty()
            }
            is org.elixir_lang.semantic.chain.For -> {
                val resolveResultList = mutableListOf<ResolveResult>()

                For.treeWalkDown(call, ResolveState.initial()) { child, _ ->
                    if (child is Call) {
                        resolveResultList.addAll(callToResolveResults(child, name, arity))
                    }

                    true
                }

                resolveResultList
            }
            else -> emptyList()
        }

    private fun nameArityIntervalToResolveResult(
        call: Call,
        name: Name,
        arity: Arity,
        nameArityInterval: NameArityInterval
    ): PsiElementResolveResult? {
        val definerName = nameArityInterval.name

        return if (definerName.startsWith(name)) {
            val definerArityInterval = nameArityInterval.arityInterval
            val validResult = (arity in definerArityInterval) && (definerName == name)

            PsiElementResolveResult(call, validResult)
        } else {
            null
        }
    }
}
